### DISCUSSION TITLE
Youtube stream quality

### DISCUSSION THREAD

**Discussion by zxctw** [streamlink/streamlink#4777] [Q&A] [answered]
I just found out that when using Streamlink to stream YT, Streamlink can only get 720p. But the YT source has 2160p50. Is this normal?

(This is true in both Linux and Windows environments. Streamlink is the latest version)

**Comment by bastimeyer** [comment, marked as answer]
The YouTube plugin does only support a subset of stream qualities, depending on the video/stream.

See https://github.com/streamlink/streamlink/issues/2936

**Reply by zxctw** [reply]
So is this a YT side issue?

**Reply by bastimeyer** [reply]
I'm not 100% well-versed on the topic, but YouTube requires signing stream URLs for certain stream types via a custom signing implementation hidden inside their obfuscated JS that's embedded on each video page. If a stream URL that requires signing is requested without the signing parameters, then the download speed will be throttled to a speed that's not suitable for playback of even the lowest available bitrate. This signing implementation is rather complex, it can change at any time, and if it doesn't get re-implemented in Python (which is already bad considering the complexity and volatility), then it requires a JS runtime and executing that code (which is much, much worse).

So far, nobody has made any efforts here re-implementing the stream URL signing in Python. Projects like yt-dlp, pytube and others have done this though or they've implemented a very much slimmed down JS parser, so that the parsed JavaScript AST can be evaluated and translated to Python to be then safely executed. Use these projects instead. Streamlink does support HLS live streams on YouTube, which is the intended purpose anyway. If you feel like working on bringing this functionality to Streamlink, please go for it and submit a pull request.

In addition to that, certain streams/videos are only available as DASH streams, and Streamlink's DASH implementation lacks certain features for being able to play them back, hence why DASH is not supported by the YouTube plugin. The supported adaptive streams are direct HTTP streams of unsigned stream URLs which are remuxed via FFmpeg, but those are only a subset of the available streams and stream combinations.

**Reply by zxctw** [reply]
OK, I get the general idea. Thank you very much for the information！

**Comment by zxctw** [comment]
@bastimeyer  Just found an old version of Streamlink on my device. After testing, found that this version can get 4K60 YT video. Don't know if this will help you guys?

![2160p60](https://user-images.githubusercontent.com/49835773/187061393-65e35d36-a22a-429f-98f9-dcf84f9bf45d.PNG)

**Reply by bastimeyer** [reply]
```
$ streamlink --loglevel=debug 'https://www.youtube.com/watch?v=0Jt8Wu1osgI'
[cli][debug] OS:         Linux-5.18.16-1-git-x86_64-with-glibc2.36
[cli][debug] Python:     3.10.6
[cli][debug] Streamlink: 4.3.0+24.g95306044
[cli][debug] Dependencies:
[cli][debug]  isodate: 0.6.1
[cli][debug]  lxml: 4.9.1
[cli][debug]  pycountry: 22.3.5
[cli][debug]  pycryptodome: 3.14.1
[cli][debug]  PySocks: 1.7.1
[cli][debug]  requests: 2.28.1
[cli][debug]  websocket-client: 1.3.2
[cli][debug] Arguments:
[cli][debug]  url=https://www.youtube.com/watch?v=0Jt8Wu1osgI
[cli][debug]  --loglevel=debug
[cli][debug]  --player=mpv
[cli][info] Found matching plugin youtube for URL https://www.youtube.com/watch?v=0Jt8Wu1osgI
[plugins.youtube][debug] Using video ID: 0Jt8Wu1osgI
[plugins.youtube][debug] MuxedStream: v 299 a 251 = 1080p60
[plugins.youtube][debug] MuxedStream: v 308 a 251 = 1440p60
[plugins.youtube][debug] MuxedStream: v 315 a 251 = 2160p60
[plugins.youtube][debug] MuxedStream: v 302 a 251 = 720p60
[plugins.youtube][debug] MuxedStream: v 135 a 251 = 480p
[plugins.youtube][debug] MuxedStream: v 133 a 251 = 240p
[plugins.youtube][debug] MuxedStream: v 160 a 251 = 144p
Available streams: audio_mp4a, audio_opus, 144p (worst), 240p, 360p, 480p, 720p, 720p60, 1080p60, 1440p60, 2160p60 (best)
```

**Reply by zxctw** [reply]
I'm also using the latest version of Streamlink, but I can't get the highest picture quality.  I have this on my router as well as in Windows environment.

![擷取](https://user-images.githubusercontent.com/49835773/187071672-5caab965-55e7-43da-b86e-ba1977a06ee6.PNG)

**Reply by bastimeyer** [reply]
You are missing FFmpeg.

https://streamlink.github.io/install.html#dependencies

> FFmpeg
> Required for [muxing](https://en.wikipedia.org/wiki/Multiplexing#Video_processing) multiple video/audio/subtitle streams into a single output stream.

https://streamlink.github.io/cli.html#cmdoption-ffmpeg-ffmpeg
https://streamlink.github.io/cli/config.html

----

https://github.com/streamlink/windows-builds#notes

> Installers
> The [ffmpeg-ffmpeg](https://streamlink.github.io/cli.html#cmdoption-ffmpeg-ffmpeg) argument in the config file always gets updated to the current install-location.

> Portable
> Since no config file will be generated either, users will need to create one themselves, including the [ffmpeg-ffmpeg](https://streamlink.github.io/cli.html#cmdoption-ffmpeg-ffmpeg) argument. Otherwise, [--ffmpeg-ffmpeg](https://streamlink.github.io/cli.html#cmdoption-ffmpeg-ffmpeg) needs to be set on the command line to enable muxed output streams.

**Reply by zxctw** [reply]
Thank you very much for your reminder, Now I can get the quality I need.
